### PR TITLE
implemented generation of main dex classes list for multi dex

### PR DESCRIPTION
**Pull request by psorobka**
I've promised some time ago that I will do better multi dex support, so here it is :)
Implementation is based on mainDexClasses script, included in build-tools 21.1.1+, but it's written in Java - there should be no problems with multi OS support (doh).

There is a sample project included with integration test: multidexsample
Basically, all you need to do is to specify in pom.xml:

``` xml
<configuration>
    ....
    <dexMultiDex>true</dexMultiDex>
    <dexGenerateMainDexList>true</dexGenerateMainDexList> 
    ....
</configuration>
```

Then add dependency:

``` xml
        <dependency>
            <groupId>com.android.support</groupId>
            <artifactId>multidex</artifactId>
            <version>1.0.1</version>
            <type>aar</type>
        </dependency>
```

If you don't have the jar in your repo, you can use bundled with Android SDK Maven repo:

``` xml
    <repositories>
        <repository>
            <id>local-android-repository</id>
            <url>file://${env.ANDROID_HOME}/extras/android/m2repository</url>
        </repository>
    </repositories>
```

Finally, use multidex in your Application class:

``` java
...
import android.support.multidex.MultiDex;
...
public class YourApplication extends Application {
...
    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        ...
        MultiDex.install(this);
    }
```

I've tested it on linux and windows box.

### CODE REVIEW

**Comment by athkalia** [comment]
Any updates on this ?

File: src/test/java/com/simpligility/maven/plugins/android/sample/AbstractSampleIT.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/*
```
Comment (mosabua) [inline src/test/java/com/simpligility/maven/plugins/android/sample/AbstractSampleIT.java]: Could you remove that class its usage. It hides too much detail and makes the tests harder to read imho.

**Comment by mosabua** [comment]
Can you rebase this and ensure all tests pass?

File: src/main/java/com/simpligility/maven/plugins/android/phase08preparepackage/DexMojo.java [hunk 2]
```diff
@@ -570,32 +595,70 @@ private static File getJavaExecutable()
...
         
-        File mainDexClasses = new File( targetDirectory, "mainDexClasses.txt" );
-        commands.add( mainDexClasses.getAbsolutePath() );
+        commands.add( "-Djava.ext.dirs=" + getAndroidSdk().getBuildToolsLibDirectoryPath() );
+        commands.add( "com.android.multidex.MainDexListBuilder" );        
+        commands.add( mainDexClassesJar.getAbsolutePath() );
         
         Set< File> inputFiles = getDexInputFiles();  
         StringBuilder sb = new StringBuilder();
-        sb.append( '"' ).append( StringUtils.join( inputFiles, File.pathSeparatorChar ) ).append( '"' );
```
Comment (mosabua) [inline src/main/java/com/simpligility/maven/plugins/android/phase08preparepackage/DexMojo.java]: From what I remember this was necessary for things to work on OSX, Linux and windows..

**Comment by sytolk** [comment]
Maybe my issue https://github.com/simpligility/android-maven-plugin/issues/556 is for the file separator difference on my OS ?

**Comment by mosabua** [comment]
I released 4.4.1 now but I would like to release another version soon and include this. Could you rebase and ensure tests path please.

**Pull request merged**